pandas/tools/tests/test_merge.py

```diff
@@ -1368,3 +1369,118 @@ def f():
+        left = DataFrame(
+            {'X': x,
+             'Y': x.cat.set_categories(['bar', 'foo', 'bah'])})
+        right = self.right.drop_duplicates(['X'])
+        right = right.assign(
+            Y=pd.Series(['foo', 'foo']).astype(
+                'category', categories=['foo', 'bar', 'baz']))
+
+        merged = pd.merge(left, right, on='X')
+        result = merged.dtypes.sort_index()
+        expected = Series([CategoricalDtype(),
+                           CategoricalDtype(),
+                           CategoricalDtype(),
+                           np.dtype('O')],
+                          index=['X', 'Y_x', 'Y_y', 'Z'])
+        assert_series_equal(result, expected)
+
+    def test_categories_same(self):
+        # GH 10409
+        left = self.left.assign(X=self.left.X.astype('category'))
+        right = self.right.assign(X=self.right.X.astype('category'))
+
+        merged = pd.merge(left, right, on='X')
+        result = merged.dtypes.sort_index()
+        expected = Series([CategoricalDtype(),
+                           np.dtype('O'),
+                           np.dtype('O')],
+                          index=['X', 'Y', 'Z'])
+        assert_series_equal(result, expected)
+
+    def test_categories_different(self):
+
+        r = self.right.drop_duplicates(['X'])
+
+        # from above with original categories
+        left = self.left.assign(X=self.left.X.astype('category'))
+
+        right = r.assign(X=r.X.astype('category'))
+        merged = pd.merge(left, right, on='X')
+
+        # swap the categories
+        # but should still work (end return categorical)
+        left = self.left.assign(X=self.left.X.astype('category'))
+        right = r.assign(X=r.X.astype('category', categories=['foo', 'bar']))
+        result = pd.merge(left, right, on='X')
+        tm.assert_index_equal(result.X.cat.categories,
+                              pd.Index(['bar', 'foo']))
+
+        assert_frame_equal(result, merged)
+
+        result = result.dtypes.sort_index()
+        expected = Series([CategoricalDtype(),
+                           np.dtype('O'),
+                           np.dtype('O')],
+                          index=['X', 'Y', 'Z'])
+        assert_series_equal(result, expected)
+
+        # swap the categories and ordered on one
+        # but should still work (end return categorical)
+        right = r.assign(X=r.X.astype('category', categories=['foo', 'bar'],
+                                      ordered=True))
+        result = pd.merge(left, right, on='X')
+        tm.assert_index_equal(result.X.cat.categories,
+                              pd.Index(['bar', 'foo']))
+
+        assert_frame_equal(result, merged)
+
+        result = result.dtypes.sort_index()
+        expected = Series([CategoricalDtype(),
+                           np.dtype('O'),
+                           np.dtype('O')],
+                          index=['X', 'Y', 'Z'])
+        assert_series_equal(result, expected)
```

Comment:
Should there be a test here added with actual different categories that results in object type of the merge column X